internal/appsec/internal/protection/waf/waf.go

```diff
@@ -0,0 +1,93 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package waf
+
+import (
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	httpinstr "gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/instrumentation/http"
+	waftypes "gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/internal/protection/waf/types"
+	appsectypes "gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/types"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/dyngo"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	"github.com/sqreen/go-libsqreen/waf"
+	"github.com/sqreen/go-libsqreen/waf/types"
+)
+
+type EventManager interface {
+	SendEvent(event *appsectypes.SecurityEvent)
+}
+
+// List of rule addresses currently supported by the WAF
+const (
+	serverRequestRawURIAddr  = "server.request.uri.raw"
+	serverRequestHeadersAddr = "server.request.headers.no_cookies"
+)
+
+// NewOperationEventListener returns the WAF event listener to register in order to enable it.
+func NewOperationEventListener(appsec EventManager) dyngo.EventListener {
+	wafRule, err := waf.NewRule(staticWAFRule)
+	if err != nil {
+		log.Error("appsec: waf error: %v", err)
+		return nil
+	}
+
+	return httpinstr.OnHandlerOperationStartListener(func(op *dyngo.Operation, args httpinstr.HandlerOperationArgs) {
+		// For this handler operation lifetime, create a WAF context and the list of detected attacks
+		var (
+			// TODO(julio): make the attack slice thread-safe once we listen for sub-operations
+			attacks []waftypes.RawAttackMetadata
+			wafCtx  = waf.NewAdditiveContext(wafRule)
+		)
+
+		httpinstr.OnHandlerOperationFinish(op, func(op *dyngo.Operation, res httpinstr.HandlerOperationRes) {
+			// Release the WAF context
+			wafCtx.Close()
+			// Log the attacks if any
+			if len(attacks) > 0 {
+				// Create the base security event out of the slide of attacks
+				event := appsectypes.NewSecurityEvent(attacks, httpinstr.MakeHTTPOperationContext(args, res))
+				// Check if a span exists
+				if span := args.Span; span != nil {
+					// Add the span context to the security event if any
+					spanCtx := span.Context()
+
+					event.AddContext(appsectypes.SpanContext{
+						TraceID: spanCtx.TraceID(),
+						SpanID:  spanCtx.SpanID(),
+					})
+					// Keep this span due to the security event
+					span.SetTag(ext.SamplingPriority, ext.ManualKeep)
+				}
```

Comment:
@knusbaum or @gbbr: everything went well when removing the data event but this block. I wanted to use `tracer.SpanFromContext()` but we (obviously?) end up having circular imports because the tracer imports appsec, which ultimately imports the WAF, which imports back the tracer. The only solution I found and is here implemented was:
1. Use `ddtrace` interface types to avoid importing `tracer`
2. Don't use `SpanFromContext()` but rather the span, directly given in a new `Span ddtrace.Span` field of `HandlerOperationArgs`.

Let me know if you think there's something better to do here 🤔